# Mutant library design
This subdirectory contains scripts and notebooks that generate primer pools used for Omicron_BA.1 variant spike deep mutational scanning experiments. The design is run by [Snakefile](Snakefile), and puts results in [./results/](results). Code generates four separate primer pools:
 - [./results/gisaid_primer_oPool.csv](./results/gisaid_primer_oPool.csv) spreadsheet contains specific forward and reverse primer pools for all mutations present in GISAID spike sequences.
 - [./results/usher_primer_oPool.csv](./results/usher_primer_oPool.csv) spreadsheet contains specific forward and reverse primers for each independently reoccurring mutation on SARS-CoV-2 phylogenetic tree.
 - [./results/positiveSel_primer_oPool.csv](./results/positiveSel_primer_oPool.csv) spreadsheet contains random forward and reverse primers for each spike site that's undergoing positive selection.
 - [./results/paired_positiveSel_primer_oPool.csv](./results/paired_positiveSel_primer_oPool.csv) spreadsheet contains forward and reverse primers for generating spike mutants that have paired mutations for closely occurring spike residues, which are undergoing positive selection.

The final primer sequences used as ordering sheets for oPools on `IDTdna` are `.csv` files ending with `oPool` in [./results/](results).

The file [results/aggregated_mutations.csv](results/aggregated_mutations.csv) indicates all mutations that are designed in each category.

## Input data
- mutation count data from [CoVsurver](https://mendel.bii.a-star.edu.sg/METHODS/corona/current/MUTATIONS/hCoV-19_Human_2019_WuhanWIV04/hcov19_Spike_mutations_table.html) was downloaded on January-31-2022.
- [./reference_sequences](./reference_sequences) subdirectory contains SARS-CoV-2 spike reference sequences and lookup tables required to renumber positions between variants and [other_plasmid_maps_for_library_design](./reference_sequences/other_plasmid_maps_for_library_design) subdirectory within it contains additional plasmid sequences used for library cloning. 

## Scripts
- [scripts](./scripts) subdirectory contains scripts for generating data. Scripts work as follows:
  - `spike_positive_selection_sites.py` script uses SARS-CoV-2 spike protein selection data (as described in [this](https://www.medrxiv.org/content/10.1101/2021.06.21.21259286v1.full) paper) to filter for positions in spike that are undergoing positive selection. 
  - `filter_and_align_gisaid.py` uses `GISAID_data/spikeprot0724.fasta` sequences to align all SARS-CoV-2 spike sequences deposited in GISAID as of July-26-2021 
  - `spike_alignment_counts.py`  extracts all mutations in GISAID spike alignments relative too Wuhan-1 sequence.
  - `spike_mutcounts.py` counts the number of independently reoccurring mutations on SARS-CoV-2 phylogenetic tree available from [UShER](https://github.com/yatisht/usher).
  - `2021Jan_create_primers.py` and `create_primers_del.py` are scripts that create random or specific amino acid change primers, respectively. 

## Notebooks
- [./notebooks](notebooks) subdirectory contains notebooks used to generate primer pools found in [./results/primers](results/primers). 
  - `gisaid_variant_primers.py.ipynb` notebook generates specific amino acid primers for each mutation present in GISAID data.
  - `usher_primers.py.ipynb` notebook generates specific amino acid primers for each independently reoccurring mutation on SARS-CoV-2 phylogenetic tree.
  - `positive_selection_primers.py.ipynb` notebook generates NNG/NNC primer pools for each position on spike that is undergoing positive selection.
  - `paired_positive_selection_primers.py.ipynb` notebook generates pools of NNG/NNC primers that introduce paired mutations for closely located sites that are undergoing positive selection
  - `oPool_primer_sheets.py.ipynb` takes primer pools generated by the notebooks above and formats spreadsheets in accordance to `IDTdna` oPool order input format.



